golem-worker-executor-base/src/durable_host/blobstore/container.rs

```diff
@@ -410,3 +422,112 @@ impl<Ctx: WorkerCtx> HostStreamObjectNames for DurableWorkerCtx<Ctx> {
 
 #[async_trait]
 impl<Ctx: WorkerCtx> Host for DurableWorkerCtx<Ctx> {}
+
+#[async_trait]
+impl<Ctx: WorkerCtx> HostContainer for &mut DurableWorkerCtx<Ctx> {
```

Comment:
I still don't know the purpose of these traits but there's a lot of boilerplate here, if possible it would be nice to avoid it (will look for other implementations that may explain).

Reply:
You might be right on the current permit system for host function calls having flaws.  It might add unnecessary boilerplate and is prone to developer errors due to the separate permit acquisition step.

Maybe we can just integrate permit acquisition into `record_host_function_call`.

This improved function could:

* Reduce redundant code.
* Guarantee permit acquisition.
* Maintain existing synchronization.
* Eliminate the possibility of forgetting to acquire a permit.